### PR TITLE
fix(HtmlParser): mark <source> elements as void

### LINKED ISSUE #5663: <source> tag throwing error since alpha.48 (d388c0a)
It appears that angular2 wants the `<source .. >` tag to be closed. <source ..></source> and <source /> are both invalid according to w3 spec (as far as I know).

Markup:
```
<video autoplay muted loop poster="assets/videos/traffic-1/traffic-1.png">
      <source src="assets/videos/traffic-1/traffic-1.webm" type="video/webm">
      <source src="assets/videos/traffic-1/traffic-1.mp4" type="video/mp4">
</video>
```

Console log:
```
EXCEPTION: Template parse errors:
Unexpected closing tag "video" ("ideo/webm">
          <source src="assets/videos/traffic-1/traffic-1.mp4" type="video/mp4">
        [ERROR ->]</video>
      </div>
      <div class="overlay">
"): e@6:8
Unexpected closing tag "div" ("       <source src="assets/videos/traffic-1/traffic-1.mp4" type="video/mp4">
        </video>
      [ERROR ->]</div>
      <div class="overlay">
        <img src="assets/logos/medium-white.png" class="m-logo"/>
"): e@7:6
```

Ref: http://www.w3.org/TR/html-markup/source.html

### PR DESCRIPTION
**Pull request by pkozlowski-opensource**
Fixes #5663

### CODE REVIEW

**Comment by pkozlowski-opensource** [comment]
@vicb picked up this one to get (a bit) familiar with the new HTML parser code. Not sure what you think about having tests for each specific tags

**Comment by vicb** [comment]
> Not sure what you think about having tests for each specific tags

Not sure if it would be really useful - if you write the source and the tests, they would probably pass. however it doesn't hurt either.

There are also more void elements that are not currently covered, could you add them ? (http://www.w3.org/TR/html5/syntax.html#void-elements)

File: modules/angular2/src/compiler/html_tags.ts [hunk 1]
```diff
@@ -369,6 +369,7 @@ var TAG_DEFINITIONS: {[key: string]: HtmlTagDefinition} = {
   'rp': new HtmlTagDefinition({closedByChildren: ['rb', 'rt', 'rtc', 'rp'], closedByParent: true}),
   'optgroup': new HtmlTagDefinition({closedByChildren: ['optgroup'], closedByParent: true}),
   'option': new HtmlTagDefinition({closedByChildren: ['option', 'optgroup'], closedByParent: true}),
+  'source': new HtmlTagDefinition({requiredParents: ['audio', 'video'], isVoid: true}),
```
Comment (vicb) [inline modules/angular2/src/compiler/html_tags.ts]: `requiredParents` means that the first parent (`audio`) will be automatically added when none of them is present. I don't think it's like that in the spec, is it ?

**Comment by pkozlowski-opensource** [comment]
@vicb updated. Mind having another look?

File: modules/angular2/test/compiler/html_parser_spec.ts [hunk 2]
```diff
@@ -75,6 +75,26 @@ export function main() {
...
         });
 
+        it('should not error on void elements from HTML5 spec', () => {
+          // http://www.w3.org/TR/html-markup/syntax.html#syntax-elements without:
+          // <base> - it can be present in head only
+          // <meta> - it can be present in head only
+          // <command> - obsolete
+          // <keygen> - obsolete
+
+          humanizeDom(parser.parse('<map><area></map>', 'TestComp'));
```
Comment (vicb) [inline modules/angular2/test/compiler/html_parser_spec.ts]:
```
[
  'fix1',
  'fix2',
].forEach((html) => {
  expect(parser.parse(html, 'TestComp').errors).toEqual([]);
});
```

File: modules/angular2/src/compiler/html_tags.ts [hunk 3]
```diff
@@ -357,7 +362,8 @@ var TAG_DEFINITIONS: {[key: string]: HtmlTagDefinition} = {
   }),
   'td': new HtmlTagDefinition({closedByChildren: ['td', 'th'], closedByParent: true}),
   'th': new HtmlTagDefinition({closedByChildren: ['td', 'th'], closedByParent: true}),
-  'col': new HtmlTagDefinition({closedByChildren: ['col'], requiredParents: ['colgroup']}),
+  'col': new HtmlTagDefinition(
+      {closedByChildren: ['col'], requiredParents: ['colgroup'], isVoid: true}),
```
Comment (vicb) [inline modules/angular2/src/compiler/html_tags.ts]: `col` is a void el, you can remove `closedByChildren: ['col'], `

File: modules/angular2/test/compiler/html_parser_spec.ts (hunk 2 above)
Comment (pkozlowski-opensource) [inline modules/angular2/test/compiler/html_parser_spec.ts]: :+1:

File: modules/angular2/test/compiler/html_parser_spec.ts [hunk 4]
```diff
@@ -75,6 +75,22 @@ export function main() {
               ]);
         });
 
+        it('should not error on void elements from HTML5 spec',
+           () => {
+
```
Comment (vicb) [inline modules/angular2/test/compiler/html_parser_spec.ts]: nits:
- remove the extra blank lines
- have one fixture per line (I think that if you add a "," after the last element clang will not mess with the formatting)
- close the `.forEach()` with a `;`

**Comment by pkozlowski-opensource** [comment]
@vicb clang-format seems to have _very_ strong opinion about how things should work.... I'm not sure I find it easier to read now. But hey, this is why we've got formatter to make us all agree :-)

Feel free to fiddle more with the formatting if you've got an idea on making it look nicer

File: modules/angular2/test/compiler/html_parser_spec.ts [hunk 5]
```diff
@@ -75,6 +75,18 @@ export function main() {
...
+           () => {// http://www.w3.org/TR/html-markup/syntax.html#syntax-elements without:
+                  // <base> - it can be present in head only
+                  // <meta> - it can be present in head only
+                  // <command> - obsolete
+                  // <keygen> - obsolete
+                  ['<map><area></map>', '<div><br></div>', '<colgroup><col></colgroup>',
+                   '<div><embed></div>', '<div><hr></div>', '<div><img></div>',
+                   '<div><input></div>', '<object><param>/<object>', '<audio><source></audio>',
+                   '<audio><track></audio>', '<p><wbr></p>',
+        ].forEach((html) => { expect(parser.parse(html, 'TestComp').errors).toEqual([]); })});
```
Comment (vicb) [inline modules/angular2/test/compiler/html_parser_spec.ts]: `});});` missing ";"

**Comment by vicb** [comment]
LGTM

**Pull request closed without merging**